### ISSUE TITLE
compilation issues after latest pull

### ISSUE THREAD

**Issue by daxav** [kokkos/kokkos#3003]
Pulled latest version out of git. Can't compile my project as I get some previously non-existing compilation errors. Is something broken here? Or significantly changed. 
I suspect it may have to do something with supporting C++17. When compiling under different rules for CUDA (without C++17) these issues do not arise.

```
In file included from ~/kokkos/core/src/Kokkos_Parallel.hpp:53:0,
~/kokkos/core/src/Kokkos_View.hpp:478:23: error: ‘Kokkos::is_always_assignable_v’ declared as an ‘inline’ variable
 inline constexpr bool is_always_assignable_v =
```

```
In file included from ~/kokkos/core/src/Kokkos_Core.hpp:58:0,
~/kokkos/core/src/Kokkos_OpenMP.hpp:231:8: error: ‘DeviceTypeTraits’ is not a class template
 struct DeviceTypeTraits<OpenMP> {
 ~/kokkos/core/src/Kokkos_OpenMP.hpp:231:33: error: explicit specialization of non-template ‘Kokkos::Profiling::Experimental::DeviceTypeTraits’
 struct DeviceTypeTraits<OpenMP> {
                                 ^
~kokkos/core/src/Kokkos_OpenMP.hpp:232:20: error: ‘DeviceType’ does not name a type
   static constexpr DeviceType id = DeviceType::OpenMP;
```

**Comment by dalg24** [comment]
2nd error reported is related to #3001 and fixed in #3004

**Comment by dalg24** [comment]
1st error message makes little sense to me.  How did you configure Kokkos?  What compiler did you use?

@dhollman it feels like a compiler issue to me.  I would prefer not rather not guard these with `__cpp_inline_variables`.

**Comment by daxav** [comment]
> 1st error message makes little sense to me. How did you configure Kokkos? What compiler did you use?

Compiling with g++ 6.3.0 20170516
Some relevant flags:
-DKOKKOS_ENABLE_CXX17 -std=c++17
KOKKOS_DEVICES = "OpenMP"
KOKKOS_OPTIONS = "aggressive_vectorization,disable_profiling"

**Comment by dalg24** [comment]
Inline variables are [only supported from GCC 7](https://en.cppreference.com/w/cpp/compiler_support#cpp17)

https://godbolt.org/z/TaLHwF

We need to guard that definition :/

**Comment by crtrott** [comment]
No we don't need to guard that - GCC6 simply doesn't support enough C++17 so you better use GCC7 if you want C++17 with Kokkos. If we really want to we can put somewhere an error out inside of cmake to say that we don't support GCC 6 with C++17 enabled. Inline variables are such a fundamental concept of C++17 not some random corner case.

**Comment by dalg24** [comment]
Then fail at configuration time?

**Comment by crtrott** [comment]
does cmake have build-in checks for C++17 features? Or just one by one stuff?

**Comment by dalg24** [comment]
I don't see any for inline variables [here](https://cmake.org/cmake/help/latest/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#prop_gbl:CMAKE_CXX_KNOWN_FEATURES)

**Comment by dalg24** [comment]
@daxav the short answer is you cannot enable C++17 with GCC-6

**Comment by dalg24** [comment]
https://github.com/kokkos/kokkos/blob/874be6a26ed1401245e263b08f26d295e77752a9/cmake/kokkos_corner_cases.cmake#L50-L53
```
$ cmake .. -DCMAKE_CXX_STANDARD=17
-- Setting default Kokkos CXX standard to 17
-- The CXX compiler identification is GNU 6.5.0
-- Check for working CXX compiler: /usr/local/bin/c++
-- Check for working CXX compiler: /usr/local/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The project name is: Kokkos
-- SERIAL backend is being turned on to ensure there is at least one Host space. To change this, you must enable another host execution space and configure with -DKokkos_ENABLE_SERIAL=OFF or change CMakeCache.txt
-- Using -std=gnu++1z for C++17 extensions as feature
-- Execution Spaces:
--     Device Parallel: NONE
--     Host Parallel: NONE
--       Host Serial: SERIAL
--
-- Architectures:
-- Found TPLLIBDL: /usr/lib/x86_64-linux-gnu/libdl.so
CMake Error at cmake/kokkos_corner_cases.cmake:52 (MESSAGE):
  You have requested c++17 support for GCC 6.5.0.  Although CMake has allowed
  this and GCC accepts -std=c++1z/c++17, GCC <= 6 does not properly support
  *this capture.  Please reduce the C++ standard to 14 or upgrade the
  compiler if you do need 17 support
Call Stack (most recent call first):
  cmake/kokkos_tribits.cmake:186 (INCLUDE)
  CMakeLists.txt:156 (KOKKOS_SETUP_BUILD_ENVIRONMENT)


-- Configuring incomplete, errors occurred!
See also "/scratch/build_gcc6/CMakeFiles/CMakeOutput.log".
```

**Comment by crtrott** [comment]
CMake would actually caught that from our side, since we disallow it inside of cmake (see @dalg24 comment).

**Comment by daxav** [comment]
> @daxav the short answer is you cannot enable C++17 with GCC-6

Yeah, that's clear, and compiling with gcc-8 works fine for the first issue. Thanks.

**Comment by daxav** [comment]
> 2nd error reported is related to #3001 and fixed in #3004

I gather this means that this has been solved in release 3.1.1? 
I've updated, but the error persists. 
```
~/kokkos/core/src/Kokkos_OpenMP.hpp:231:8: error: ‘DeviceTypeTraits’ is not a class template
 struct DeviceTypeTraits<OpenMP> {
        ^~~~~~~~~~~~~~~~
~/kokkos/core/src/Kokkos_OpenMP.hpp:231:33: error: explicit specialization of non-template ‘Kokkos::Profiling::Experimental::DeviceTypeTraits’
 struct DeviceTypeTraits<OpenMP> {
                                 ^
~/kokkos/core/src/Kokkos_OpenMP.hpp:232:20: error: ‘DeviceType’ does not name a type; did you mean ‘Device’?
   static constexpr DeviceType id = DeviceType::OpenMP;
                    ^~~~~~~~~~
                    Device
```

Am I missing something?

**Comment by masterleinad** [comment]
No, the header was also missing at another place, see #3012.

**Comment by crtrott** [comment]
:-(

**Comment by daxav** [comment]
I have now updated with #3084. If I understand properly, changes from #3012 have been approved, and therefore, this issue should also be solved.
However, I am still seeing errors (see below). Is this actually solved, and I am missing something? Or is it still a standing issue?

`In file included from ~/kokkos/core/src/Kokkos_Core.hpp:58,
                 from ~/kokkos/core/src/impl/Kokkos_ExecPolicy.cpp:45:
~/kokkos/core/src/Kokkos_OpenMP.hpp:231:8: error: ‘DeviceTypeTraits’ is not a class template
 struct DeviceTypeTraits<OpenMP> {
        ^~~~~~~~~~~~~~~~
~/kokkos/core/src/Kokkos_OpenMP.hpp:231:33: error: explicit specialization of non-template ‘Kokkos::Profiling::Experimental::DeviceTypeTraits’
 struct DeviceTypeTraits<OpenMP> {
                                 ^
~/kokkos/core/src/Kokkos_OpenMP.hpp:232:20: error: ‘DeviceType’ does not name a type; did you mean ‘Device’?
   static constexpr DeviceType id = DeviceType::OpenMP;`

**Comment by DavidPoliakoff** [comment]
@daxav could we get your current configure/cmake line to try to reproduce?

**Comment by daxav** [comment]
I don't really have a cmake configuration on my project. It is built by a Makefile, which includes Makefile.kokkos. 

Relevant compilation flags are -DKOKKOS_ENABLE_CXX17 -std=c++17
KOKKOS_DEVICES = "OpenMP"
KOKKOS_OPTIONS = "aggressive_vectorization,disable_profiling"

Don't know if this actually helps @DavidPoliakoff. 
Would the issue be solved by actually following the recommendation of setting up kokkos as a package and building my project integrally through cmake? I did not set this up in my current project, so I am not sure why it has been done like this.

**Comment by crtrott** [comment]
We will actually remove the disable_profiling option. You can still disable the LIBDL library dependency (which was implied by that option) but the profiling hooks will always be available.

**Comment by ndellingwood** [comment]
Merged PR #3235 should fix profiling issues, sets `KOKKOS_ENABLE_LIBDL`
